src/app/tests/TestReadInteraction.cpp

```diff
@@ -1813,60 +1782,60 @@ void TestReadInteraction::TestSubscribeUrgentWildcardEvent(nlTestSuite * apSuite
         nonUrgentDelegate.mGotEventResponse = false;
         nonUrgentDelegate.mGotReport        = false;
 
-        // wait for min interval 2 seconds (in test, we use 1.6 seconds considering the time variation), expect no event is
+        // wait for min interval 1 seconds (in test, we use 0.6 seconds considering the time variation), expect no event is
         // received, then wait for 0.8 seconds, then the urgent event would be sent out
         //  currently DriveIOUntil will call `DriveIO` at least once, which means that if there is any CPU scheduling issues,
         // there's a chance 1.9s will already have elapsed by the time we get there, which will result in DriveIO being called when
         // it shouldn't. Better fix could happen inside DriveIOUntil, not sure the sideeffect there.
+
+        // Advance monotonic looping to allow events to trigger
         while (true)
         {
-            if ((System::SystemClock().GetMonotonicTimestamp() - startTime) >= System::Clock::Milliseconds32(600))
+            ctx.GetIOContext().DriveIO();
+            if ((mockClock.GetMonotonicTimestamp() - startTime) >= System::Clock::Milliseconds32(600))
             {
                 break;
             }
-            ctx.GetIOContext().DriveIO(); // at least one IO loop is guaranteed
+            mockClock.AdvanceMonotonic(System::Clock::Milliseconds32(10));
```

Comment:
With the mock clock, why is spinning still needed? What's the difference between advancing by 10ms, and advancing once in one go? What exactly of the implementation is causing you to have to do that? Is the IO you expected not occurring?

Reply:
The usual approach is to have a `DriveIOUntilIdle()`. Is there a way to implement this with existing DriveIO ?

Reply:
There was a comment about this from the previous designer about using DriveIOUntil, whic explained why he had to rely on loops in the first place as DriveIOUntil would be suceptible to scheduling issues.

In the meantime, it relies on the SystemClock to timeout, which won't work due to usage of a mock clock.

I can confirm I had the same problems when I tried to use it.

We also have DrainAndServiceIO but there again the function description states:

- ... This should run to completion
     * in well-behaved logic (i.e there isn't an indefinite ping-pong of messages transmitted back
     * and forth).

And there again I can confirm that in our case it doesn't work as we do have back and forth.